solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/main/java/com/solace/spring/cloud/stream/binder/util/SolaceAsyncMessageProducer.java

```diff
@@ -0,0 +1,126 @@
+package com.solace.spring.cloud.stream.binder.util;
+
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+import com.solace.spring.cloud.stream.binder.messaging.CorrelationData;
+import com.solacesystems.jcsmp.Destination;
+import com.solacesystems.jcsmp.JCSMPException;
+import com.solacesystems.jcsmp.JCSMPProperties;
+import com.solacesystems.jcsmp.JCSMPSession;
+import com.solacesystems.jcsmp.JCSMPStreamingPublishCorrelatingEventHandler;
+import com.solacesystems.jcsmp.XMLMessage;
+import com.solacesystems.jcsmp.XMLMessageProducer;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+public class SolaceAsyncMessageProducer {
+
+	private static final Log logger = LogFactory.getLog(SolaceAsyncMessageProducer.class);
+	private static final ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);
+
+	private final Map<String, CorrelationData> pendingReturns = new ConcurrentHashMap<>();
+	private final XMLMessageProducer producer;
+	private Consumer<Exception> errorHandler;
+
+	SolaceAsyncMessageProducer(JCSMPSession jcsmpSession) throws JCSMPException {
+		this.producer = jcsmpSession.getMessageProducer(publisherEventHandler);
+
+		final int pubAckTimeOut = getPubAckTimeOut(jcsmpSession);
+
+		// mem leak protection.
+		scheduler.scheduleAtFixedRate(
```

Comment:
How does the broker/jcsmp lib handel following situations:
- The session si not connected. How long will the jcsmp lib try to send the message?
- The jcsmp lib transmitted to the broker, the broker is not HA and has a power lost restart. Will the lib resend the message after broker restarted or is the message lost without any JCSMPStreamingPublishCorrelatingEventHandler methode?
- Any further edge cases?

Reply:
It all depends on the reconnect-retries property value. If the API is still trying to reconnect to the broker, the API will retransmit queued AD messages after reconnect. When the API aborts a reconnect attempt, it stops retransmitting messages.

The broker restart is a DR case. The API will renumber and retransmit AD messages after reconnect in DR cases. Note that in a DR case, a message could be published as a duplicate. So it is up to applications to check for duplicates.

Otherwise everything should be fine.

Reply:
I agree and changed the way it works. I have had to play around a little more with the correlationKey to see how it works.